lib/gis/omp_threads.c

```diff
@@ -0,0 +1,52 @@
+#if defined(_OPENMP)
+#include <omp.h>
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <grass/gis.h>
+#include <grass/glocale.h>
+
+/*! \brief A helper function to setup the number of threads for C modules
+   supported by OpenMP
+   \param opt A nprocs Option struct to specify the number of threads
+   \return the number of threads set up for OpenMP parallel computing
+ */
+
+int G_set_omp_num_threads(struct Option *opt)
+{
+    /* make sure the nproc Option is given */
+    if (opt == NULL)
+        G_fatal_error(_("Option is NULL."));
+    else if (opt->key == NULL)
+        G_fatal_error(_("Option key is NULL."));
+    else if (strcmp(opt->key, "nprocs") != 0)
+        G_fatal_error(_("This function can only be used for 'nprocs' Option."));
+
+    int threads = atoi(opt->answer);
+#if defined(_OPENMP)
+    int CPU_threads = omp_get_num_procs();
+    if (threads > CPU_threads) {
+        G_warning(_("The number of threads specified for OpenMP is more "
```

Comment:
The advantage of using `struct Option` is that you can include the name (key) in the messages:

```suggestion
        G_warning(_("The number of threads specified in %s for OpenMP (%d) is more "
```

...and then use `threads, opt->key`.